src/Microsoft.Health.Dicom.Core/Features/Partitioning/PartitionService.cs

```diff
@@ -25,15 +28,31 @@ public PartitionService(PartitionCache partitionCache, IPartitionStore partition
         _logger = EnsureArg.IsNotNull(logger, nameof(logger));
     }
 
-    public async Task<AddPartitionResponse> AddPartitionAsync(string partitionName, CancellationToken cancellationToken = default)
+    public async Task<GetOrAddPartitionResponse> GetOrAddPartitionAsync(string partitionName, CancellationToken cancellationToken = default)
     {
         EnsureArg.IsNotNull(partitionName, nameof(partitionName));
 
-        _logger.LogInformation("Creating partition with name '{PartitionName}'.", partitionName);
+        _logger.LogInformation("Getting partition with name '{PartitionName}'.", partitionName);
+
+        PartitionNameValidator.Validate(partitionName);
+
+        Partition partition = await _partitionCache.GetAsync(partitionName, partitionName, _partitionStore.GetPartitionAsync, cancellationToken);
```

Comment:
Would it be simpler if the `_partitionStore` had a `GetOrAddAsync` like [`ConcurrentDictionary<TKey, TValue>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=net-7.0)?

Reply:
Simpler in the long term yes, but I'm not sure the complexity of that change in this PR is worth the effort since the goal was just to move the existing logic to a place that could be used in other processes like functions.

Reply:
I'd feel a lot more comfortable if we did this work sooner than later to ensure our more concurrent operations are done at the source, like adding a new sproc. If we don't do it now, I'd like to see it done very soon haha